core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -235,47 +302,88 @@ protected MapDiff() {
             adds = new HashMap<>();
         }
 
-        protected MapDiff(StreamInput in, KeyedReader<T> reader) throws IOException {
+        public abstract void writeKey(K key, StreamOutput out) throws IOException;
+
+        public abstract K readKey(StreamInput in) throws IOException;
+
+        protected MapDiff(StreamInput in, KeyedReader<K, T> reader) throws IOException {
             deletes = new ArrayList<>();
             diffs = new HashMap<>();
             adds = new HashMap<>();
             int deletesCount = in.readVInt();
             for (int i = 0; i < deletesCount; i++) {
-                deletes.add(in.readString());
+                deletes.add(readKey(in));
             }
 
             int diffsCount = in.readVInt();
             for (int i = 0; i < diffsCount; i++) {
-                String key = in.readString();
+                K key = readKey(in);
                 Diff<T> diff = reader.readDiffFrom(in, key);
                 diffs.put(key, diff);
             }
 
             int addsCount = in.readVInt();
             for (int i = 0; i < addsCount; i++) {
-                String key = in.readString();
+                K key = readKey(in);
                 T part = reader.readFrom(in, key);
                 adds.put(key, part);
             }
         }
+
         @Override
         public void writeTo(StreamOutput out) throws IOException {
             out.writeVInt(deletes.size());
-            for (String delete : deletes) {
-                out.writeString(delete);
+            for (K delete : deletes) {
+                writeKey(delete, out);
             }
 
             out.writeVInt(diffs.size());
-            for (Map.Entry<String, Diff<T>> entry : diffs.entrySet()) {
-                out.writeString(entry.getKey());
+            for (Map.Entry<K, Diff<T>> entry : diffs.entrySet()) {
+                writeKey(entry.getKey(), out);
                 entry.getValue().writeTo(out);
             }
 
             out.writeVInt(adds.size());
-            for (Map.Entry<String, T> entry : adds.entrySet()) {
-                out.writeString(entry.getKey());
+            for (Map.Entry<K, T> entry : adds.entrySet()) {
+                writeKey(entry.getKey(), out);
                 entry.getValue().writeTo(out);
             }
         }
     }
+
+    private static abstract class StringKeyedMapDiff<T extends Diffable<T>, M> extends MapDiff<String, T, M> {
```

Comment:
I wonder if we should follow the OpenIntMap and call this StringMapDiff , just to key things shorter.